### PR TITLE
Changed K nearest neighbor search to use kd trees

### PR DESCRIPTION
**Pull request by sveerara**
Edited NearestNodeLocator and SlaveNeighborhoodThread to use kd-tree to calculate the k nearest neighbors for each slave node. 

Refs #9246

### CODE REVIEW

File: framework/src/geomsearch/NearestNodeLocator.C
Comment (permcody) [inline framework/src/geomsearch/NearestNodeLocator.C]: This will need work for DistributedMesh

File: framework/src/geomsearch/NearestNodeLocator.C
Comment (andrsd) [inline framework/src/geomsearch/NearestNodeLocator.C]: Why cannot you use the local variable?

File: framework/src/geomsearch/SlaveNeighborhoodThread.C
Comment (permcody) [inline framework/src/geomsearch/SlaveNeighborhoodThread.C]: std::size_t instead of size_t. In C++ it's in a namespace, you are picking up the C version.

File: framework/src/geomsearch/SlaveNeighborhoodThread.C
Comment (andrsd) [inline framework/src/geomsearch/SlaveNeighborhoodThread.C]: This is probably ok for right now. Ideally `KdTree` would be a MOOSE class that would have `ret_index`, `ret_dist_sqr` and similar encapsulated with the `kd_tree_t` and nice API that would take `Node`s/`Point`s etc.

This code would be a lot more readable.

File: framework/src/geomsearch/SlaveNeighborhoodThread.C
Comment (permcody) [inline framework/src/geomsearch/SlaveNeighborhoodThread.C]: Not safe with DistributedMesh. Either we'll need to work that out, or we should protect this class from being using with DistributedMesh.

File: framework/src/geomsearch/SlaveNeighborhoodThread.C
Comment (permcody) [inline framework/src/geomsearch/SlaveNeighborhoodThread.C]: Don't use C arrays - C++ has std::array for this kind of thing.

**Review by andrsd — DISMISSED** [review]

**Review by permcody — CHANGES_REQUESTED** [review]
This is a nice enhancement. Just need to go through and modernize all your code and follow code standards.

File: framework/include/geomsearch/NearestNodeLocator.h
Comment (andrsd) [inline framework/include/geomsearch/NearestNodeLocator.h]: I had the same comment, but I think that nanoflann expects this method in the templated class that users are supposed to write. Note that she does not call it anywhere, so I think it gets called from inside the nanoflann magic.

Comment (permcody) [inline framework/include/geomsearch/NearestNodeLocator.h]: Makes sense, let's at least document why it doesn't match guidelines then.

File: framework/src/geomsearch/SlaveNeighborhoodThread.C
Comment (andrsd) [inline framework/src/geomsearch/SlaveNeighborhoodThread.C]: I think it is again the nanoflann's API. We could hide this behind our API as I suggested in my review. I do not like this either. If you hate this, then let's have a `KDTree` class in MOOSE that will hide all of this ugliness.

File: framework/src/geomsearch/SlaveNeighborhoodThread.C
Comment (sveerara) [inline framework/src/geomsearch/SlaveNeighborhoodThread.C]: After talking with @bwspenc, it looks like the contact code is not yet ready to be used with distributed mesh. I will try to add an error to make sure this does not run with distributed mesh.

**Comment by sveerara** [comment]
@permcody and @andrsd, thanks for your comments on the pull request. I will make the syntax changes and resubmit the commit.

File: framework/src/geomsearch/SlaveNeighborhoodThread.C
Comment (sveerara) [inline framework/src/geomsearch/SlaveNeighborhoodThread.C]: So should I create a KDTree class in MOOSE or leave this as it is for now?

File: framework/src/geomsearch/NearestNodeLocator.C
Comment (sveerara) [inline framework/src/geomsearch/NearestNodeLocator.C]: This is the max leaf size for the kd_tree. Should the user be allowed to set this parameter with the default value being 10?

Comment (permcody) [inline framework/src/geomsearch/NearestNodeLocator.C]: The problem isn't the value, it's the fact that it's an unnamed constant hardcoded in the code. A better approach would be to make it a static constant member of the class or even a local scope constant in the implementation file with a documentation string describing it's usage.

File: framework/src/geomsearch/SlaveNeighborhoodThread.C
Comment (permcody) [inline framework/src/geomsearch/SlaveNeighborhoodThread.C]: Use this method:
`errorIfDistributedMesh("<ClassName>")`

File: framework/src/geomsearch/SlaveNeighborhoodThread.C
Comment (permcody) [inline framework/src/geomsearch/SlaveNeighborhoodThread.C]: Either way, if you leave it as is though, we'll want to see documentation as to why you are deviating from the standard.

File: framework/src/geomsearch/NearestNodeLocator.C
Comment (andrsd) [inline framework/src/geomsearch/NearestNodeLocator.C]: > This is the max leaf size for the kd_tree. Should the user be allowed to set this parameter with the default value being 10?

That is a way how to fix the magic constant if you want to give users that power.

File: framework/src/geomsearch/SlaveNeighborhoodThread.C
Comment (andrsd) [inline framework/src/geomsearch/SlaveNeighborhoodThread.C]: I am leaning towards the `KDTree` class. If I did not miss something, it can live outside `NearestNodeLocator` and I would say, put it in `utils/`. 

Also comments like "Need to convert this into a C-array because that's what nanoflann expects" will help explain things like this to people not familiar with the interface.

**Comment by bwspenc** [comment]
@sveerara says she's getting errors like this when she tries to run 'git clang-format':
```
YAML:25:15: error: unknown key 'SortIncludes'
SortIncludes: false
              ^~~~~
Error reading /Users/veers/projects/moose_contact/moose/.clang-format: Invalid argument
clang-format did not modify any files
```
Any ideas?

**Comment by permcody** [comment]
Tag @rwcarlsen.

**Comment by sveerara** [comment]
I have applied the patch manually for now to pass the precheck test. The version of clang I am using is "/opt/moose/llvm-3.7.0/bin/clang-format" which is two versions older than what @bwspenc is using. Could the error be because of the older version?

**Comment by andrsd** [comment]
You need to use clang 3.9...

**Comment by sveerara** [comment]
@andrsd, thanks, I will install the newer moose package.

@permcody and @andrsd , I have created a KDTree MOOSE class in utils and moved the nanoflann parts to that file.

File: framework/include/utils/KDTree.h
Comment (permcody) [inline framework/include/utils/KDTree.h]: 👍

File: framework/include/utils/KDTree.h
Comment (permcody) [inline framework/include/utils/KDTree.h]: If size has to bee the same as LIBMESH_DIM, we probably don't need a separate parameter that we are passing around. Maybe this is a good candidate for a "static assert" How would this fail?

File: framework/include/utils/KDTree.h
Comment (permcody) [inline framework/include/utils/KDTree.h]: We aren't big fans of this keyword because it has been the root-cause of long-standing (and I mean long) bugs in our codes that cause problems for months or years. The problem is that it gives you a free pass to have data that can change in an object when it's declared constant. This can be especially bad with threading. Is this keyword needed here? I don't see where you are creating a constant version of this object.

File: framework/include/utils/KDTree.h
Comment (permcody) [inline framework/include/utils/KDTree.h]: Can use the new alias syntax

File: framework/include/utils/KDTree.h
Comment (permcody) [inline framework/include/utils/KDTree.h]: Looks like this hasn't been implemented? At least not according to the docs.

File: framework/src/geomsearch/NearestNodeLocator.C
Comment (permcody) [inline framework/src/geomsearch/NearestNodeLocator.C]: With the way you are constructing this object then calling build tree, are you sure you can't snag the dynamic memory you need all at once in the constructor?

File: framework/src/utils/KDTree.C
Comment (permcody) [inline framework/src/utils/KDTree.C]: Use `libmesh_make_unique<T>(args)` here instead. That's superior to using `new` which can fail under extraordinary circumstances. Then you shouldn't need to explicitly assert below. It'll abort the program if the memory allocation fails.

Comment (permcody) [inline framework/src/utils/KDTree.C]: Background `libmesh_make_unique` just calls `std::make_unique<T>` but since that's C++14, it's not always there so it falls back to a safe implementation on some compilers.

File: framework/src/utils/KDTree.C
Comment (permcody) [inline framework/src/utils/KDTree.C]: You can put the default value in as a second argument at construction and drop the explicit loop below.

File: framework/src/utils/KDTree.C
Comment (permcody) [inline framework/src/utils/KDTree.C]: Thanks for the comment but this is actually a good thing. All indexing in C++ is std::size_t but people routinely (wrongly) use `unsigned int`. (Like every single loop counter everywhere, e.g. `unsigned int i = 0; i < v.size(); ++i`) Below you have another comment about turning std::size_t into an unsigned int with an extra vector. Why not just return and use the std::size_t everywhere? This is the default for all indexing operations. Even `vector::size()` returns a std::size_t but people store it into an unsigned int all the time. Yes everybody sucks at getting types right. 😉  Just use size_t here and don't worry about converting back and forth to unsigned int.

File: framework/src/utils/KDTree.C
Comment (permcody) [inline framework/src/utils/KDTree.C]: Yeah, I don't understand why we need this. We should probably just return std::size_t and use it everywhere.

File: framework/src/utils/KDTree.C
Comment (permcody) [inline framework/src/utils/KDTree.C]: This is some sort of tuning parameter? Do you know much about it or what value it should be? Should it be constant at least here? Maybe even a class static const?

**Review by permcody — CHANGES_REQUESTED** [review]
Looking pretty good, I had a few questions and a few more suggestions though.

**Comment by friedmud** [comment]
Do you have any demonstrations that this is faster?

The NearestNodeNeighborhoods can be perfectly tailored to the problem at hand (I've even used weird shapes and fuzzy bounding boxes to shape them before).

Not only that - but they are ordered in such a way as to find a match quickly... and setup to optimally allow for rebuilding.

Not dissing this work... just... there is a lot more thought and work in the NearestNodeNeighborhood stuff than you might know... and I want to make sure that this new stuff is just as fast / efficient.

**Comment by sveerara** [comment]
@friedmud, I tried a simple mechanical contact problem with a 2D block sitting on another 2D block of the same dimension.

When the slave and master boundaries both have 100 nodes, the performance log for NearestNodeLocator::findNodes() says that the existing method takes 0.30 % of Active time while the kd tree method takes 0.26% of Active time. The findNodes() function gets called 11 times during the simulation but the patch is probably not updated in all 11 calls. 

I tried with 1000 nodes and timing is 0.18% with the existing approach and 0.13% with kd trees. With 2000 nodes, the timing is 0.16% with the existing approach and 0.11% with the kd tree approach.

So overall I think the Kdtree approach is a little faster than the existing approach. I have not tried this approach on weird shapes but I don't see a reason why it would not work for any shape. If you have a particular example where you think this might not work, I am happy to try it out before this gets merged. Also, I have not made any changes to the bounding box part of the NearestNodeLocator code. The Kd tree approach does work well on the existing contact verification problems. 

The reason we are trying to make the geometric search faster is that we would like to update the patch more often. In some of the mechanical contact problems, once a node moves out of the patch, contact is not detected and this leads to erroneous results. So if mechanical contact is not detected for a slave node within the existing neighbor patch, we would like to update the patch for that particular node and then check for contact again.

**Comment by bwspenc** [comment]
@sveerara Do you have the times in seconds for the search with the two approaches?

This is obviously a pretty small part of the overall run time, but as @sveerara said, the reason we want this to be as fast as possible is that the next step here is that we want to re-run the search for nodes that are outside the patch.  The goal here is to make it so that a user will never encounter a case where contact isn't enforced when it should be because the node slid out of the patch.  The options that were added a couple of years ago to dynamically update the patch are a huge help, but we are still running into issues with it, especially when we slide past a lot of faces in a single time step.

**Comment by friedmud** [comment]
At 0.16% of the runtime - why are we even doing this?

Can you create a case that shows that any of this is a limiting factor in a
run?

**Comment by permcody** [comment]
Agreed - Let's go ahead and build the case where you update the patch a lot
more and lets see if we need this.

**Comment by permcody** [comment]
@sveerara - Any progress on this? I know you are busy this week but this has been sitting for over a month now.

**Comment by sveerara** [comment]
@permcody, I made some code changes during weekend to check the timings for the case where the geometric search gets run many more times. I am running some test cases today. Hopefully I will have some timing comparison between the kd-tree and priority queue approach by today or tomorrow.

**Comment by friedmud** [comment]
Before this gets merged I would really like to talk about it and look at
the implementation during the tiger team. I just want to better understand
the implications of this change.

**Comment by sveerara** [comment]
@friedmud, that sounds good. Once I run the timing comparisons, I will add the changed code in which the geometric search can be switched between priority queue approach and kd-tree approach using a flag.

**Comment by sveerara** [comment]
@friedmud, I tried a 2D sliding block contact problem with 2,000 elements on slave boundary and 30,000 elements on the master boundary. As the top block slides on the bottom block, some of the slave nodes will fall outside the patch that was created at the first time step. So the patch of K closest master nodes is updated for just these slave nodes. If the Jacobian is not updated during the patch update using Fande's PR #9509, then the actual time spent in contact search using either the priority queue or Kd-tree is less than 1% of the total simulation time. 85% of the time was spent in calculating Jacobian.

When I pulled in Fande's code and added only dofs connected to the slave node and the closest master node to the Jacobian (which can be updated every nonlinear iteration), the search time becomes more significant (patch update takes 0.11% of total time using Kd-Tree and priority queue approach takes 10% of total time). Note that I used the default patch size (I guess 40) for this comparison. From this example, I think it is beneficial to switch from prioirty queue to Kd-tree because it will help at a later stage when all these other code changes get merged in. 

The code changes for the automatic patch update and Jacobian editing, and the sliding block example problem are in https://github.com/sveerara/moose/tree/contact_patch. priority_queue flag can be used to switch between priority queue and Kd-tree. automatic_patch_update flag can be used to turn on the automatic patch update and update_jacobian_preallocation flag can be used to edit the Jacobian entries dynamically. More changes are required to make these code changes work in parallel but the code at this stage is mostly to compare speed between the two different contact search approaches. Also, I tried these example problems with petsc-3.7.6. I am not sure if it works with any other petsc version.

**Comment by permcody** [comment]
@sveerara - We're going to allow this to go in. Please address comments and repush and we'll get it reviewed.

**Comment by sveerara** [comment]
@permcody Great! Thanks. I will address your comments soon and repush the merge request.

File: framework/include/utils/KDTree.h
Comment (sveerara) [inline framework/include/utils/KDTree.h]: This function is required by the nanoflann library. If the return value is false, the standard bbox computation is used. Else the bbox computation written within this computation is used.

File: framework/include/utils/KDTree.h
Comment (sveerara) [inline framework/include/utils/KDTree.h]: I changed the syntax to `using KdTreeT = nanoflann::`. I am not sure if this is the new syntax.

File: framework/include/utils/KDTree.h
Comment (sveerara) [inline framework/include/utils/KDTree.h]: I have removed the mutable keyword.

File: framework/src/geomsearch/NearestNodeLocator.C
Comment (sveerara) [inline framework/src/geomsearch/NearestNodeLocator.C]: I have moved the building tree part to the constructor as you suggested and it works fine.

File: framework/src/utils/KDTree.C
Comment (sveerara) [inline framework/src/utils/KDTree.C]: The Kdtree basically divides the master point cloud into bins based on the distance. This parameter specifies the maximum number of points that can be present in a bin. As the max_leaf_size increases, the time taken to  build the tree decreases but the time taken to conduct the search increases. I don't know if there is an optimal value that works well for all problems.

I took this value from the examples within libmesh/contrib/nanoflann. I could add this as another parameter (like patch size) in MooseMesh with default value of 10 and users can change it if needed. Would that be okay?

File: framework/src/utils/KDTree.C
Comment (sveerara) [inline framework/src/utils/KDTree.C]: Thanks for the information. I have changed `new` to `libmesh_make_unique<T>(args)`.

File: framework/src/utils/KDTree.C
Comment (sveerara) [inline framework/src/utils/KDTree.C]: Thanks for the comment. I have changed the code to return a vector of std::size_t instead of unsigned int.

File: framework/include/utils/KDTree.h
Comment (sveerara) [inline framework/include/utils/KDTree.h]: Yes, size has to be same as LIBMESH_DIM. I have removed the extra parameter as you suggested.

**Comment by sveerara** [comment]
Should I use RunException for the distributed mesh tests that are failing?

**Comment by permcody** [comment]
No - you should restrict the test so that it doesn't run with DistributedMesh in your test specification. We'll definitely want to fix this sooner than later though. We are pushing to get DistributedMesh working with most if not all of our physics this year.

**Comment by permcody** [comment]
@sveerara - You might retest with DistributedMesh once more after the libMesh update. Several things have been fixed.

**Comment by sveerara** [comment]
@permcody I removed  ` _mesh.errorIfDistributedMesh("NearestNodeLocator");` and the tests seem to be working fine with distributed mesh.

File: framework/include/utils/KDTree.h
Comment (permcody) [inline framework/include/utils/KDTree.h]: 👍

File: framework/src/utils/KDTree.C
Comment (permcody) [inline framework/src/utils/KDTree.C]: C++11 supports this. You can use `std::array<3, Real>`, then when passing data to the C API, you use `.data()` which gives you access to the first entry of the "array". It may not be worth changing now, but you don't need raw arrays.

**Review by permcody — APPROVED** [review]

**Comment by permcody** [comment]
@andrsd, rather than dismiss your review, I'll let you take another look and do it yourself. Thanks

**Pull request merged**
